### PR TITLE
Bug 1616931 - support logging via tc-worker-runner

### PR DESCRIPTION
**Pull request by djmitche**
This does not implement all of the functionality, but is a self-contained series of commits that gets partway along the journey.

 * Establish a new `log` message type and capability
 * Build a simple logging abstraction that supports both structured and unstructured logging
 * Pipe all logs from the worker, and all logs from worker-runner, through that abstraction
 * Update generic-worker to use `log` messages too.

The result is pretty similar behavior to the existing behavior: all logs end up going to worker-runner's stderr, with golang-style timestamp prefixes

What remains:
 * Support for backends other than stdio
    * syslog, for sending directly to papertrail without involving the system logger (I haven't found a golang library for this yet..)
    * stackdriver? some other structured log aggregator?
 * docker-worker support (using the `log` message type)
 * Add structured logs to workers where necessary (@walac maybe the WORKER_METRICS makes sense to do with structured logging?)

Bugzilla Bug: [1616931](https://bugzilla.mozilla.org/show_bug.cgi?id=1616931)

### CODE REVIEW

**Comment by petemoore** [comment]
Be aware, there seem to be some failures: https://community-tc.services.mozilla.com/tasks/groups/BF2_ixaBQ7m0ZWsJ_ya-mQ

File: tools/taskcluster-worker-runner/runner/run.go [hunk 1]
```diff
@@ -165,6 +166,13 @@ func Run(configFile string) (state run.State, err error) {
 	// set up protocol
 
 	proto := protocol.NewProtocol(transp)
+
+	// Register to receive log messages on the given protocol
+	proto.Register("log", func(msg protocol.Message) {
+		logging.Destination.LogStructured(msg.Properties)
+	})
```
[the remark this answers is not in the document]
Comment (djmitche) [reply tools/taskcluster-worker-runner/runner/run.go]: It's listed in `knownCapabilities`.  I think this is the same issue as in https://github.com/taskcluster/taskcluster/pull/2425#discussion_r387192726 so let's solve it there first :)

[the remark this answers is not in the document]
Comment (walac) [reply tools/taskcluster-worker-runner/runner/run.go]: I mean, removing the `knownCapabilities` and adding it here.

[the remark this answers is not in the document]
Comment (djmitche) [reply tools/taskcluster-worker-runner/runner/run.go]: I'll rebase on top of your work and do that :)

**Comment by djmitche** [comment]
> Be aware, there seem to be some failures: https://community-tc.services.mozilla.com/tasks/groups/BF2_ixaBQ7m0ZWsJ_ya-mQ

I don't know why that happened -- go.mod isn't modified in this patch.  It seems OK now.

I think getting `go mod tidy` into `yarn generate` will help - it's very easy to know that when the `yarn generate` gets a red X, just run `yarn generate` :)

**Comment by djmitche** [comment]
Waiting on #2501 now :)

**Comment by djmitche** [comment]
OK!  I'd love another look from both of you.  This is rebased on recent protocol changes.  The "replace StdioTransport with PipeTransport" commit is pretty big, but most of that is testing support.  I'd be interseted in feedback on the protocol/testing package, but feel free to ignore the `_test.go` files if you'd like.

**Comment by djmitche** [comment]
Those three red X's are on the unsupported platforms that have since been disabled in master :)

**Comment by djmitche** [comment]
rebased :)

**Comment by djmitche** [comment]
Green!

**Pull request merged**